tests/e2e/tx/service_test.go

```diff
@@ -899,6 +916,169 @@ func (s IntegrationTestSuite) TestTxDecode_GRPCGateway() {
 
 			res, err := testutil.PostRequest(fmt.Sprintf("%s/cosmos/tx/v1beta1/decode", val.APIAddress), "application/json", req)
 			s.Require().NoError(err)
+			if tc.expErr {
+				s.Require().Contains(string(res), tc.expErrMsg)
+			} else {
+				var result tx.TxDecodeResponse
+				err := val.ClientCtx.Codec.UnmarshalJSON(res, &result)
+				s.Require().NoError(err)
+
+				txb := authtx.WrapTx(result.Tx)
+				tx, err := val.ClientCtx.TxConfig.TxEncoder()(txb.GetTx())
+				s.Require().NoError(err)
+				s.Require().Equal(encodedTxBytes, tx)
+			}
+		})
+	}
+}
+
+func (s IntegrationTestSuite) TestTxEncodeAmino_GRPC() {
+	val := s.network.Validators[0]
+	txBuilder := s.mkTxBuilder()
+	stdTx, err := clienttx.ConvertTxToStdTx(val.ClientCtx.LegacyAmino, txBuilder.GetTx())
+	s.Require().NoError(err)
+	txJSONBytes, err := val.ClientCtx.LegacyAmino.MarshalJSON(stdTx)
+	s.Require().NoError(err)
+
+	testCases := []struct {
+		name      string
+		req       *tx.TxEncodeAminoRequest
+		expErr    bool
+		expErrMsg string
+	}{
+		{"nil request", nil, true, "request cannot be nil"},
+		{"empty request", &tx.TxEncodeAminoRequest{}, true, "invalid empty tx json"},
+		{"invalid request", &tx.TxEncodeAminoRequest{AminoJson: "invalid tx json"}, true, "invalid request"},
+		{"valid request with amino-json", &tx.TxEncodeAminoRequest{AminoJson: string(txJSONBytes)}, false, ""},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		s.Run(tc.name, func() {
+			res, err := s.queryClient.TxEncodeAmino(context.Background(), tc.req)
+			if tc.expErr {
+				s.Require().Error(err)
+				s.Require().Contains(err.Error(), tc.expErrMsg)
+				s.Require().Empty(res)
+			} else {
+				s.Require().NoError(err)
+				s.Require().NotEmpty(res.GetAminoBinary())
+
+				stdTxConfig := legacytx.StdTxConfig{Cdc: val.ClientCtx.LegacyAmino}
+				newStdTx, err := stdTxConfig.TxJSONDecoder()(res.GetAminoBinary())
```

Comment:
The correct logic is :
- you either unmarshal the AminoBinary bytes
- or you unmarshalJSON the AminoJSON bytes

Here you're unmarshallingJSON the AminoBinary bytes, which should be a red flag.

Reply:
I'm a bit confused initially and it makes more sense now, thanks for the clarification amaury